### PR TITLE
Implement a few more diagnostics in SwiftParser

### PR DESCRIPTION
**Pull request by ahoppen**
A few unrelated commits to migrate more diagnostics from the C++ parser to SwiftParser.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Tests/SwiftParserTest/translated/GenericDisambiguationTests.swift [hunk 1]
```diff
@@ -237,8 +237,7 @@ final class GenericDisambiguationTests: XCTestCase {
   func testGenericDisambiguation15() {
     AssertParse(
       """
-      // TODO: parse empty <> list
-      //A<>.c() // e/xpected-error{{xxx}}
+      A<>.c()
```
Comment (bnbarham) [inline Tests/SwiftParserTest/translated/GenericDisambiguationTests.swift:240]: What tree does this produce?

Comment (ahoppen) [reply Tests/SwiftParserTest/translated/GenericDisambiguationTests.swift:240]: Thanks for checking. It parses `<>` as an operator. I’ll leave this TODO in for now.

File: Tests/SwiftParserTest/translated/HashbangLibraryTests.swift [hunk 2]
```diff
@@ -22,10 +22,7 @@ final class HashbangLibraryTests: XCTestCase {
       class Foo {}
       // Check that we diagnose and skip the hashbang at the beginning of the file
       // when compiling in library mode.
-      """,
-      diagnostics: [
-        // TODO: Old parser expected error on line 1: hashbang line is allowed only in the main file
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Tests/SwiftParserTest/translated/HashbangLibraryTests.swift:27]: We parse the shebang as trivia. I would like to parse it as part of the source file at some point (rdar://98309158) but the current implementation isn’t too bad.

Added a substructure test.

File: Sources/SwiftParser/Nominals.swift [hunk 3]
```diff
@@ -150,7 +150,7 @@ extension Parser {
     introucerHandle: RecoveryConsumptionHandle
   ) -> T where T: NominalTypeDeclarationTrait {
     let (unexpectedBeforeIntroducerKeyword, introducerKeyword) = self.eat(introucerHandle)
-    let (unexpectedBeforeName, name) = self.expectIdentifier(keywordRecovery: true)
+    let (unexpectedBeforeName, name) = self.expectIdentifier(allowIdentifierLikeKeywords: false, keywordRecovery: true)
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftParser/Nominals.swift:153]: Absolutely!

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Tests/SwiftParserTest/translated/GenericDisambiguationTests.swift (hunk 1 above)
Comment (bnbarham) [reply Tests/SwiftParserTest/translated/GenericDisambiguationTests.swift:230]: Would be good to keep a TODO here (or did you open a bug?)

File: Sources/SwiftParser/Nominals.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (bnbarham) [reply Sources/SwiftParser/Nominals.swift:153]: Hey we can totally do this now :D. But fine not to for this PR.

File: Tests/SwiftParserTest/translated/HashbangLibraryTests.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (bnbarham) [reply Tests/SwiftParserTest/translated/HashbangLibraryTests.swift:27]: FWIW you don't need all those `nil`'s any more. Maybe we should update the dump to not add them either 🤔

**Review by bnbarham — APPROVED** [review]

File: Tests/SwiftParserTest/translated/GenericDisambiguationTests.swift (hunk 1 above)
Comment (ahoppen) [reply Tests/SwiftParserTest/translated/GenericDisambiguationTests.swift:230]: I wonder where I lost that change to keep the TODO. It’s in again now.

File: Tests/SwiftParserTest/translated/HashbangLibraryTests.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahoppen) [reply Tests/SwiftParserTest/translated/HashbangLibraryTests.swift:27]: Thanks, the PR is so old it still required the `nil`s when I created it. I removed them.

File: Sources/SwiftParser/Nominals.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftParser/Nominals.swift:153]: Maybe in a follow-up PR.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Pull request merged**